core/src/main/java/com/yandex/yoctodb/util/mutable/impl/FoldedByteArrayIndexedList.java

```diff
@@ -0,0 +1,179 @@
+package com.yandex.yoctodb.util.mutable.impl;
+
+import com.google.common.primitives.Ints;
+import com.google.common.primitives.Longs;
+import com.yandex.yoctodb.util.OutputStreamWritable;
+import com.yandex.yoctodb.util.UnsignedByteArray;
+import com.yandex.yoctodb.util.mutable.ByteArrayIndexedList;
+import net.jcip.annotations.NotThreadSafe;
+import org.jetbrains.annotations.NotNull;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.util.ArrayList;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.SortedMap;
+import java.util.TreeMap;
+
+/**
+ * {@link ByteArrayIndexedList} with variable sized elements
+ *
+ * @author irenkamalova
+ */
+@NotThreadSafe
+public final class FoldedByteArrayIndexedList
+        implements ByteArrayIndexedList {
+    @NotNull
+    private final Set<UnsignedByteArray> elements;
+    @NotNull
+    private final List<Long> offsets;
+    @NotNull
+    private final SortedMap<Integer, Integer> docIdOffsetIndex;
+    private final int sizeOfIndexOffsetValue; // how many bites
+    private final int databaseDocumentsCount;
+
+
+    public FoldedByteArrayIndexedList(
+            @NotNull final Map<UnsignedByteArray, LinkedList<Integer>> elements,
+            int databaseDocumentsCount) {
+
+        this.databaseDocumentsCount = databaseDocumentsCount;
+        this.docIdOffsetIndex = new TreeMap<>();
+        this.offsets = new ArrayList<>();
+        this.elements = elements.keySet();
+
+        long elementOffset = 0;
+        // reserve first value for empty documents
+        offsets.add(-1L);
+
+        for (Map.Entry<UnsignedByteArray, LinkedList<Integer>> elem :
+                elements.entrySet()) {
+            UnsignedByteArray value = elem.getKey();
+            offsets.add(elementOffset);
+
+            final long currentElementOffset = elementOffset;
+            elem.getValue().forEach( docId ->
+                    docIdOffsetIndex
+                         .put(docId,
+                         offsets.indexOf(currentElementOffset)));
```

Comment:
`O(n)` hidden in `indexOf()` is too expensive. And the result doesn't change for `docId`.
We can just add and increment `offsetIndex` to use it instead of `indexOf()`, right?